pkg/universal-cnf/config/composite_test.go

```diff
@@ -0,0 +1,282 @@
+	ipamDefaultPrefixPool        string = ""
+	ipamServerAddress            string = ""
+	ipamPrefixLength             int    = 0
+
+	// fields of type connection.Connection
+	connId             string = ""
+	connNetworkService string = "uncf"
+	machanismCls       string = ""
+	mechanismType      string = "MEMIF"
+	paraDescription    string = ""
+	paraName           string = "test"
+	paraNetnsInode     string = ""
+	paraSocketfile            = paraName + "/memif.sock"
+	srcIP              string = "192.168.22.2/30"
+	dstIP              string = "192.168.22.1/30"
+	srcRequired        bool   = true
+	dstRequired        bool   = true
+	dstRoutePrefix     string = "192.168.0.0/16"
+	labelsNamespace    string = "default"
+	ucnfPodName        string = "helloworld-ucnf-"
+	labelsPodName             = ucnfPodName + "test"
+	controlPlaneName   string = "-control-plane"
+	clusterName        string = "kind-2"
+	pathSegmentName           = clusterName + controlPlaneName
+)
+
+var (
+	// fields of type nseconfig.Endpoint
+	ucnfeLabels    = map[string]string{}
+	ipamRoute      = []string{""}
+	vl3NameServers = []string{""}
+	vl3DNSZones    = []string{""}
+)
+
+// the mock implementer of the interface UniversalCNFBackend
+type MockUcnfBackendProcessDPConfig struct {
+}
+
+/*
+these are the implementations of the ucnf endpoint backend interface, and these mock functions are used for testing
+the ProcessDPconfig function call
+*/
+func (m *MockUcnfBackendProcessDPConfig) NewDPConfig() *vpp.ConfigData {
+	return &vpp.ConfigData{}
+}
+
+func (m *MockUcnfBackendProcessDPConfig) NewUniversalCNFBackend() error {
+	return nil
+}
+
+func (m *MockUcnfBackendProcessDPConfig) ProcessClient(dpconfig interface{}, ifName string, conn *connection.Connection) error {
+	return nil
+}
+
+func (m *MockUcnfBackendProcessDPConfig) ProcessEndpoint(dpconfig interface{}, serviceName, ifName string, conn *connection.Connection) error {
+	return nil
+
+}
+
+func (m *MockUcnfBackendProcessDPConfig) ProcessDPConfig(dpconfig interface{}, update bool) error {
+	return fmt.Errorf("failed to run ProcessDPConfig() method")
+}
+
+// mock implementor
+type MockUcnfBackendReturnNil struct {
+}
+
+/*
+these are the mock functions that will return nil only, which means that the Request() and Close() functions will
+run till the end without getting errors from ProcessEndpoint and ProcessDPConfig function calls
+*/
+func (m *MockUcnfBackendReturnNil) NewDPConfig() *vpp.ConfigData {
+	return &vpp.ConfigData{}
+}
+
+func (m *MockUcnfBackendReturnNil) NewUniversalCNFBackend() error {
+	return nil
+}
+
+func (m *MockUcnfBackendReturnNil) ProcessClient(dpconfig interface{}, ifName string, conn *connection.Connection) error {
+	return nil
+}
+
+func (m *MockUcnfBackendReturnNil) ProcessEndpoint(dpconfig interface{}, serviceName, ifName string, conn *connection.Connection) error {
+	return nil
+}
+
+func (m *MockUcnfBackendReturnNil) ProcessDPConfig(dpconfig interface{}, update bool) error {
```

Comment:
It might be simpler and more flexible to be able to inject different behavior of these methods via fields defined as funcs:

```go
type MockUcnfBackendProcessDPConfig struct {
    //...
    processDPConfig func(dpconfig interface{}, update bool) error
}
//...
func (m *MockUcnfBackendProcessDPConfig) ProcessDPConfig(dpconfig interface{}, update bool) error {
    if m.processDPConfig == nil {
        return nil
    }
    return m.processDPConfig(dpconfig, update)
}
```

This way you will get the "return nil" behavior by default when creating `MockUcnfBackendProcessDPConfig ` and make it possible to override the implementation by simply setting the field value:

```go
mocked := &MockUcnfBackendProcessDPConfig{
    processDPConfig: func(dpconfig interface{}, update bool) error {
        return fmt.Errorf("failed to run ProcessDPConfig() method")
    },
}
```

Reply:
Hello Ondrej, thank you very much for the suggestion, I will make modifications.